### PR TITLE
Factor out request time zone

### PR DESCRIPTION
**Pull request by alxdarksage**
Separated out the request time zone into a separate field (rather than reading endsOn).

### CODE REVIEW

File: test/org/sagebionetworks/bridge/models/schedules/ScheduleContextTest.java [hunk 1]
```diff
@@ -67,6 +72,7 @@ public void builderWorks() {
                 .withClientInfo(clientInfo)
                 .withStudyIdentifier(studyId)
                 .withInitialTimeZone(PST)
+                .withRequestTimeZone(PST)
```
Comment (DwayneJengSage) [inline test/org/sagebionetworks/bridge/models/schedules/ScheduleContextTest.java]: We set the requestTimeZone, but we don't verify it anywhere?

Also, is it worth making requestTimeZone and initialTimeZone different to make sure we're not crossing our wires?

File: app/org/sagebionetworks/bridge/models/schedules/ScheduleContext.java [hunk 2]
```diff
@@ -200,13 +228,22 @@ public Builder withContext(ScheduleContext context) {
         }
         
         public ScheduleContext build() {
-            // pretty much everything else is optional. I would like healthCode to be required, but it's not:
-            // we use these selection criteria to select subpopulations on sign up.
-            if (now == null) {
-                now = (initialTimeZone == null) ? DateTime.now() : DateTime.now(initialTimeZone);
+            // ScheduleContext is validated to have the correct values:
+            // - requestTimeZone is required
```
Comment (DwayneJengSage) [inline app/org/sagebionetworks/bridge/models/schedules/ScheduleContext.java]: Clarify in comment that requestTimeZone is required, but is enforced in the validator, not here.

Also, is it worth having a unit test where requestTimeZone is null, since nothing in this class actually prevents that.

File: app/org/sagebionetworks/bridge/models/schedules/ScheduleContext.java [hunk 3]
```diff
@@ -200,13 +228,22 @@ public Builder withContext(ScheduleContext context) {
...
+            // - now is advisable, but still not required
+            // - endsOn OR daysAhead needs to be supplied to calculate endsOn from now
+            if (initialTimeZone == null) {
+                initialTimeZone = requestTimeZone;
+            }
```
Comment (DwayneJengSage) [inline app/org/sagebionetworks/bridge/models/schedules/ScheduleContext.java:237]: Unit test for setting initialTimeZone based off requestTimeZone?

File: app/org/sagebionetworks/bridge/models/schedules/ScheduleContext.java [hunk 4]
```diff
@@ -200,13 +228,22 @@ public Builder withContext(ScheduleContext context) {
...
+            if (now == null && requestTimeZone != null) {
+                now = DateTime.now(requestTimeZone);
+            }
```
Comment (DwayneJengSage) [inline app/org/sagebionetworks/bridge/models/schedules/ScheduleContext.java]: Before, not specifying timezone would result in now using the default timezone. Now, now is null. Is that intentional?

File: test/org/sagebionetworks/bridge/models/schedules/IntervalActivitySchedulerTest.java [hunk 5]
```diff
@@ -70,18 +70,19 @@ public void onceWithoutTimesUsesLocalTime() throws Exception {
...
         schedule.setScheduleType(ScheduleType.ONCE);
         schedule.setExpires("P2M");
+        DateTimeZone zone = DateTimeZone.forOffsetHours(-7);
         
         ScheduleContext context= new ScheduleContext.Builder()
                 .withStudyIdentifier(TEST_STUDY)
-                .withInitialTimeZone(DateTimeZone.forOffsetHours(-7))
-                .withEndsOn(ENROLLMENT.plusDays(2)) // in UTC
+                .withRequestTimeZone(zone)
+                .withDaysAhead(2)
```
Comment (DwayneJengSage) [inline test/org/sagebionetworks/bridge/models/schedules/IntervalActivitySchedulerTest.java]: In this test, the initial and the requested timezone were different. Now they are the same. This changes the semantics of the test. Are you sure we're not losing test coverage on this?

Semantically, we should be replacing endsOn.getTimeZone() with requestTimeZone, not initialTimeZone. This seems to be happening a lot in a lot of the other tests as well.

File: test/org/sagebionetworks/bridge/services/ScheduledActivityServiceRecurringTest.java [hunk 6]
```diff
@@ -179,32 +179,25 @@ public void persistedActivitiesAreFilteredByEndsOn() throws Exception {
...
-            .withEndsOn(endsOn.withHourOfDay(23).withMinuteOfHour(59).withSecondOfMinute(59))
-            .withHealthCode(testUser.getHealthCode())
-            .withUserId(testUser.getId()).build();
-    }
-}
+                .withStudyIdentifier(study.getStudyIdentifier())
+                .withClientInfo(ClientInfo.UNKNOWN_CLIENT)
+                .withInitialTimeZone(initialTimeZone)
+                .withRequestTimeZone(requestTimeZone)
+                .withNow(now.withZone(requestTimeZone))
```
Comment (DwayneJengSage) [inline test/org/sagebionetworks/bridge/services/ScheduledActivityServiceRecurringTest.java:198]: Does any of our functionality depend on the timezone of now? This change changes the timezone of now in some tests, which may change the functionality and test coverage.

File: app/org/sagebionetworks/bridge/models/schedules/ScheduleContext.java [hunk 7]
```diff
@@ -200,13 +228,22 @@ public Builder withContext(ScheduleContext context) {
...
+            if (endsOn == null && now != null) {
+                endsOn = now.plusDays(daysAhead).withHourOfDay(23).withMinuteOfHour(59).withSecondOfMinute(59)
+                        .withMillisOfSecond(999);
```
Comment (DwayneJengSage) [inline app/org/sagebionetworks/bridge/models/schedules/ScheduleContext.java]: now and requestedTimeZone may be different. This may lead you into weird states where endsOn isn't in the requestedTimeZone.

**Review by DwayneJengSage — CHANGES_REQUESTED** [review]

File: app/org/sagebionetworks/bridge/models/schedules/ScheduleContext.java (hunk 3 above)
Comment (alxdarksage) [reply app/org/sagebionetworks/bridge/models/schedules/ScheduleContext.java:237]: Added (also verifying null works).

File: test/org/sagebionetworks/bridge/services/ScheduledActivityServiceRecurringTest.java (hunk 6 above)
Comment (alxdarksage) [reply test/org/sagebionetworks/bridge/services/ScheduledActivityServiceRecurringTest.java:198]: Nothing breaks...  logically I'd expect this since we're trying to return tasks in the request time zone, so starting calculations from that time zone for scheduledOn times should work. However, I could go back to generating this in the default time zone (probably UTC in prod) if you think I should.

File: app/org/sagebionetworks/bridge/models/schedules/ScheduleContext.java (hunk 3 above)
Comment (alxdarksage) [reply app/org/sagebionetworks/bridge/models/schedules/ScheduleContext.java:237]: Added (and test for requestTimeZone == null).

**Review by alxdarksage — COMMENTED** [review]
I've updated. I didn't change the conversion of "now" to the user's request time zone. I believe this works, but I could revert it to system time zone. I was trying to set the stage to provide now through the API at a later date, if desired for testing or any other purpose.

File: app/org/sagebionetworks/bridge/models/schedules/ScheduleContext.java [hunk 8]
```diff
@@ -200,13 +228,28 @@ public Builder withContext(ScheduleContext context) {
...
+            // - endsOn OR daysAhead needs to be supplied to calculate endsOn from now
+            if (initialTimeZone == null) {
+                initialTimeZone = requestTimeZone;
+            }
+            if (requestTimeZone != null) {
+                if (now == null) {
+                    now = DateTime.now(requestTimeZone);
+                } else {
+                    now = now.withZone(requestTimeZone);    
+                }
```
Comment (DwayneJengSage) [inline app/org/sagebionetworks/bridge/models/schedules/ScheduleContext.java:243]: There are a few places in the code where we assume ScheduleContext.getNow() is not null. Under the old code, the builder would always ensure now is not null. Under the new code, we cannot guarantee that to be the case. Do we need to fix this?

File: test/org/sagebionetworks/bridge/dynamodb/DynamoScheduledActivityDaoMockTest.java [hunk 9]
```diff
@@ -131,7 +131,7 @@ public void testGetActivities() throws Exception {
         ScheduleContext context = new ScheduleContext.Builder()
             .withStudyIdentifier(TEST_STUDY)
             .withClientInfo(ClientInfo.UNKNOWN_CLIENT)
-            .withInitialTimeZone(PACIFIC_TIME_ZONE)
+            .withRequestTimeZone(PACIFIC_TIME_ZONE)
```
Comment (DwayneJengSage) [inline test/org/sagebionetworks/bridge/dynamodb/DynamoScheduledActivityDaoMockTest.java:134]: There are a lot of tests where we're replacing initialTimeZone with requestTimeZone instead of endsOn.getTimeZone() with requestTimeZone. This changes the semantics of a lot of tests, and means we could be losing test coverage. Are we sure we want to do this?

File: test/org/sagebionetworks/bridge/services/ScheduledActivityServiceRecurringTest.java (hunk 6 above)
Comment (DwayneJengSage) [reply test/org/sagebionetworks/bridge/services/ScheduledActivityServiceRecurringTest.java:198]: Tests not failing isn't the same as nothing breaking, especially if we changed the tests, as is the case here. Instead of trying to logically deduce why these tests don't matter, I'd rather keep these tests intact and prove that everything is still working.
